kratos/processes/variational_distance_calculation_process.h

```diff
@@ -185,10 +187,8 @@ class VariationalDistanceCalculationProcess : public Process
     /// Destructor.
     ~VariationalDistanceCalculationProcess() override
     {
-
-        Model& current_model = mr_base_model_part.GetModel();
-        if(current_model.HasModelPart( mAuxModelPartName ))
-            current_model.DeleteModelPart( mAuxModelPartName );
+        if(mrModel.HasModelPart( mAuxModelPartName ))
+            mrModel.DeleteModelPart( mAuxModelPartName );
```

Comment:
And why don't we create the member distance model part at construction time and we save a member reference to it? Then it will be automatically destroyed when the process is destroyed.

Reply:
Would it be destroyed? I might be missing something but I do not think so, only model can destroy it..

Reply:
Well, the model part would remain in there. You would destroy the member reference var.
The model part itself would be destroyed once you destroy de model. If you need to destroy it to free memory but keep the model we can call the clear in the process destructor.

Reply:
Ok, make sense to me. I will call the DeleteModelPart from the clear then, is that ok?

Reply:
Not at all... (this is basically the same) My idea is to avoid having a reference to the `Mode`l but one to the model part of interest.

This can be done by creating the model part inside the constructor scope and would avoid the destructor issue we have right now. Hence, instead of calling the `DeleteModelPart`, the reference is destroyed when the utility is destroyed.

The question is, do we need to explicitly  delete the model part (not a reference to it) in here? Why it can't be destroyed when the model part is destroyed?

Reply:
uh oh
box of Pandora incoming :D

remember the whole discussion abt `Model` holding and being able to delete `ModelParts` to save memory?

some said this might backfire...here we are :/

Reply:
Hi @philbucher, I told you I had an issue with nodes being missing in MPI when I use apply_outlet_process? This was the exact reason. Eventhough this is creating a duplicate model part using same nodes, I was destructing the model part, so It was removing the nodes from the OriginModelPart which made it to give division by zero errors for number of nodes.

I am using a workaround to have this process to be destructed at the very end of the simulation now.

But I have a small question, I didnt get this problem in OpenMP, only visible in MPI.

The exact question is like this. When I use VariationalDistanceCalculation instance in a local method in MPI, once the instance goes out of scope it does remove nodes from the SubModelParts of distance calculation process model part, and removes nodes from OriginalModelPart as well. Is this the intended behaviour? And Why this only happens in MPI?

Reply:
@sunethwarna I don't think that removing the nodes in the original ModelPart is intended behavior :/
No clue why it only happens in MPI
I thought the intrusive ptr would only be removed once all counters are zero

@RiccardoRossi guess this si a question for you

Reply:
I think Im not getting your point. Variational redistance creates its own "redistance_model_part" to do the computation. You want to remove this model part when you are done with redistance to free the memory. Question is then, who should do it? As it was created by variational_redistance, it makes sense that this process removes it, either in the clear or in the destructor. Afaik, analysis stage is not calling processes Clear, so I would keep the removal of the model part in the destructor.

Reply:
> @sunethwarna I don't think that removing the nodes in the original ModelPart is intended behavior :/
> No clue why it only happens in MPI
> I thought the intrusive ptr would only be removed once all counters are zero
> 
> @RiccardoRossi guess this si a question for you

It doesn't remove nodes, but it removes nodes from the OriginalModelParts submodel parts' local mesh in MPI :/

Reply:
> I think Im not getting your point. Variational redistance creates its own "redistance_model_part" to do the computation. You want to remove this model part when you are done with redistance to free the memory. Question is then, who should do it? As it was created by variational_redistance, it makes sense that this process removes it, either in the clear or in the destructor. Afaik, analysis stage is not calling processes Clear, so I would keep the removal of the model part in the destructor.

This was exactly my point, do we really need to remove the model part to free the memory or we live with that until the model is destroyed?

** Option 1**
Assuming that you require the variational distance to be computed each time your level se evolves, which most probably is each time step, destroying it in the constructor or destroying it when the model is destroyed is more or less equivalent for me.

**Option 2**
If we assume that the variational distance can be computed once, it might have sense to free this memory. For this we have two options:
- **Option 2.1**: Destroy the complete model part by doing `Model.DeleteModelPart`. This implies having a reference to the model.
- **Option 2.2**: Add a `Clear` method that can be optionally called. This method would remove all the model part entities to keep an empty model part (whose memory consumption is negligible). Note that this empty model part would be destroyed once the model is destroyed. I personally would prefer this one, as it would avoid weird errors as the one we are experiencing and setting up the model part if remeshing is required becomes more natural.

Reply:
I wished I also lived in a "single-stage" world, however we have Filling+Packing+Solidification+Warpage + ... ;) so we need to clear things when moving from one stage to the other. Im fine with option 2.2, but could this Clear method be called from destructor?

Reply:
> I wished I also lived in a "single-stage" world, however we have Filling+Packing+Solidification+Warpage + ... ;) so we need to clear things when moving from one stage to the other. Im fine with option 2.2, but could this Clear method be called from destructor?

(I had you multistage in mind when I was thinking on this :smile:) 
Yes, I'd call the `Clear` in the constructor. To sum up:
- Store a reference to the `AuxiliaryModelPart` rather than a reference to the `Model` (which is no longer needed as the `AuxiliaryModelPart` is created during contruction.
- Instead of deleting the model part, remove all its entities to free the memory.

Reply:
@KratosMultiphysics/technical-committee I consider this very important, we should discuss this and find a good solution!

Reply:
i am not sure if i follow the conversation at 100%, but the intended use of DeleteModelPart is to COMPLETELY delete it. INTENTIONALLY you cannot keep it alive after destroying it...

Reply:
guys, the problem here is THE MODEL not the model_part.

if you destroy the model then everything will be destroyed. And the model is in principe the only thing that should live for the entire simulation, surviving the end of a stage (note that the model is passed to the stage in the constructor exactly for this reason)

Reply:
I think we are having a misunderstanding here. The original problem is that in the destructor, variational resistance called:
`Model& current_model = mr_base_model_part.GetModel();`
This assumes `**mr_base_model_part**` is still "alive" at the moment redistance destructor is called, which is quite error-prone. That's why I wanted to keep a reference to the Model so we can directly call it instead of doing `mr_base_model_part.GetModel()`. This is the origin of the discussion.

Reply:
ok, got you now.

so the idea is to keep a reference to mrModel (careful, NOT a shared pointer) instead of doing
the GetModel()

Reply:
> ok, got you now.
> 
> so the idea is to keep a reference to mrModel (careful, NOT a shared pointer) instead of doing
> the GetModel()

No. The idea is to just keep a reference to the auxiliary model part in which the variational distance process is applied. Then the reference to the model is no longer needed.

Reply:
After a brief discussion with @RiccardoRossi we thing it might be better to keep the current behavior, that is the `AuxiliaryModelPart` is **destroyed** in the destructor rather than removing its entities.

Having said this, it's safer to keep a reference to the `Model` as it might happen that the `AuxiliaryModelPart` has been already removed when the `VariationalDistanceProcess` destructor is called.

Reply:
to me deleting the `ModelPart` also seems more consistent
before the `Model` this would have been a local `ModelPart` that would be removed if the process gets destructed